README.md
# fixedstring
a templated fixed-length array of `char`s, compatible with `@safe`, `@nogc`, and `nothrow` code.

# example
```d
void main() @safe @nogc nothrow
{
	FixedString!14 foo = "clang";
	foo[0] = 'd';
	foo ~= " is cool";
	assert(foo == "dlang is cool");

	foo.length = 9;

	immutable bar = fixedString!"neat";
	assert(foo ~ bar == "dlang is neat");

	// wchars and dchars are also supported
	assert(FixedString!(5, wchar)("áéíóú") == "áéíóú");

	// in fact, any type is:
	immutable int[4] intArray = [1, 2, 3, 4];
	assert(FixedString!(5, int)(intArray) == intArray);
}
```

read [the documentation](https://fixedstring.dpldocs.info/fixedstring.html) for more info!

# licence
AGPL-3.0 or later


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Fixed
- a link in this changelog was missing.

## [2.0.1] - 2022-05-24
### Fixed
- an incorrect directory structure was preventing the dub package from building.
- some syntax in this changelog had been left out.

## [2.0.0] - 2022-05-22
### Added
- this changelog! 🎉
- a new range interface - a _random access range with slicing_ - has been added that more closely follows standard practice, replacing the previous implementation that confused ranges with containers.
 - for more information on ranges, see [std.range.primitives](https://dlang.org/phobos/std_range_primitives.html).
- the previously existing unittests have been cleaned up and added to the documentation.

### Changed
- `opIndex()` now returns a range rather than a slice.
 - to compare `foo[]` with an array of elements, use `std.algorithm.comparison.equal` instead of `==`.
- gc using, throwing, inpure, and `@system` - using types are now permitted as element types. note that using such types will make the `FixedString` use those attributes accordingly; you cannot have a `FixedString!(N, SomeClass)` and expect it to work in `@nogc` code.
- the `FixedString!"string"` helper template has been changed to `fixedString!"string"`, so as to conform with the standard D style.

### Removed
- `empty()`, `front()`, and `popFront()` have been removed. use the new range interface instead.
- it is no longer possible to modify out-of-bounds elements with `opIndex`, even if the maximum size of the `FixedString` would allow it. change `length` first before modifying, or use the `~=` operator.

### Fixed
- `opAssign`, `opOpAssign`, `opEquals`, `concat` and `opBinary` all assumed they would only ever be used with the `char` element type. this oversight has been fixed.
- `opSlice` is now bounds checked.
- now compatible with dip1000.

## [1.2.0] - 2022-04-12
### Changed
- when concatenating with the `~` operator, the nearest minimum power of two is used rather than the total size. this helps with template bloat. if you need the old behaviour, use `foo.concat!(foo.size + bar.size)(bar)` instead.

### Fixed
- documentation should now show up correctly in parsers.
- several lint and coverage fixes.

## [1.1.0] - 2022-01-12
### Added
- a shorter syntax (`FixedString!"string"`)
- you can now use fixedstring with any type, not just char. note that non-char types do not have a custom toString implementation
- a small sampling of examples has been added to the readme.

## [1.0.0] - 2022-01-10
- initial release.

[Unreleased]: https://github.com/Moth-Tolias/fixedstring/compare/v2.0.1...HEAD
[2.0.1]: https://github.com/Moth-Tolias/fixedstring/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/Moth-Tolias/fixedstring/compare/v1.2.0...v2.0.0
[1.2.0]: https://github.com/Moth-Tolias/fixedstring/compare/v1.0.0...v1.2.0
[1.1.0]: https://github.com/Moth-Tolias/fixedstring/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/Moth-Tolias/fixedstring/releases/tag/v1.0.0


source/fixedstring/package.d
/**
* a @safe, @nogc-compatible template string type.
* Authors: Susan
* Date: 2021-11-06
* Licence: AGPL-3.0 or later
* Copyright: Susan, 2021
*/
module fixedstring;

import fixedstring.fixedstringrange;
import fixedstring.opapplymixin;
import std.traits: isSomeChar;

/// short syntax.
auto fixedString(string s)()
{
	import std.math.algebraic: nextPow2;
	return FixedString!(nextPow2(s.length))(s);
}

///
@safe @nogc nothrow unittest
{
	enum testString = "dlang is good";
	immutable foo = fixedString!(testString);
	immutable bar = FixedString!16(testString);

	assert(foo == bar);
}

///
struct FixedString(size_t maxSize, CharT = char)
{
	invariant (length_ <= data.length);

	enum size = maxSize; ///

	private size_t length_;
	private CharT[maxSize] data = ' ';

	///
	size_t length() const @nogc nothrow pure @safe
	{
		return length_;
	}

	/// ditto
	void length(in size_t rhs) @nogc nothrow pure @safe
	in (rhs <= maxSize)
	{
		if (rhs >= length)
		{
			data[length .. rhs] = CharT.init;
		}
		length_ = rhs;
	}

	/// constructor
	this(in CharT[] rhs)
	in (rhs.length <= maxSize)
	{
		length = rhs.length;
		data[0 .. length] = rhs[];
	}

	/// assignment
	void opAssign(in CharT[] rhs)
	in (rhs.length <= maxSize)
	{
		length = rhs.length;
		data[0 .. length] = rhs[];
	}

	/// ditto
	void opAssign(T: CharT, size_t n)(in FixedString!(n, T) rhs)
	in (rhs.length <= maxSize)
	{
		length = rhs.length;
		data[0 .. length] = rhs[0 .. length];
	}

	/// ditto
	void opOpAssign(string op)(in CharT[] rhs)
	if (op == "~")
	in (length + rhs.length <= maxSize)
	{
		immutable oldLength = length;
		length = length + rhs.length;
		data[oldLength .. length] = rhs[];
	}

	/// ditto
	void opOpAssign(string op)(in CharT rhs)
	if (op == "~")
	in (length + 1 <= maxSize)
	{
		length = length + 1;
		data[length - 1] = rhs;
	}

	/// ditto
	void opOpAssign(string op, T: CharT, size_t n)(in FixedString!(n, T) rhs)
	if (op == "~")
	in (length + rhs.length <= maxSize)
	{
		immutable oldLength = length;
		length = length + rhs.length;
		data[oldLength .. length] = rhs[0 .. rhs.length];
	}

	/// array features...
	auto opSlice(in size_t first, in size_t last) const @nogc nothrow pure @safe
	in (first <= length && last <= length)
	{
		return data[first .. last];
	}

	/// ditto
	size_t opDollar(size_t pos)() const @nogc nothrow pure @safe
	if (pos == 0)
	{
		return length;
	}

	/// ditto
	auto opIndex() const
	{
		return FixedStringRange!CharT(data[0 .. length]);
	}

	/// ditto
	CharT opIndex(in size_t index) const @nogc nothrow pure @safe
	in (index < length)
	{
		return data[index];
	}

	/// ditto
	void opIndexAssign(in CharT rhs, in size_t index)
	in (index < length)
	{
		data[index] = rhs;
	}

	/// equality
	bool opEquals(T : CharT, size_t n)(in FixedString!(n, T) rhs) const
	{
		import std.algorithm.comparison: equal;
		return this[].equal(rhs[]);
	}

	/// ditto
	bool opEquals(in CharT[] s) const
	{
		if (length != s.length)
		{
			return false;
		}

		import std.algorithm.comparison: equal;
		return this[].equal(s[]);
	}

	/// concatenation. note that you should probably use the ~ operator instead - only use this version when you are pressed for ram and aren't making many calls, or you will end up with template bloat.
	auto concat(size_t s, T: CharT, size_t n)(in FixedString!(n, T) rhs) const
	in (s >= length + rhs.length)
	{
		FixedString!(s) result;

		result = this;
		result ~= rhs;

		return result;
	}

	/// concatenation operator. generally, you should prefer this version.
	auto opBinary(string op, T: CharT, size_t n)(in FixedString!(n, T) rhs) const
	if (op == "~")
	{
		import std.math.algebraic: nextPow2;
		return concat!(nextPow2(size + rhs.size))(rhs);
	}

	static if (isSomeChar!CharT)
	{
		///
		const(CharT)[] toString() const nothrow pure @safe
		{
			return data[0 .. length].idup;
		}
	}

	///
	size_t toHash() const @nogc nothrow pure @safe
	{
		ulong result = length;
		foreach (CharT c; data[0 .. length])
		{
			result += c;
		}
		return result;
	}

	mixin(opApplyWorkaround);
}

///
@safe @nogc nothrow pure unittest
{
	immutable string temp = "cool";
	auto foo = FixedString!8(temp);
	assert(foo[0] == 'c');
	assert(foo == "cool");

	import std.algorithm.comparison: equal;
	assert(foo[].equal("cool"));
	assert(foo[0 .. $] == "cool");

	foo[2] = 'd';
	assert(foo == "codl");
	assert(foo != "");

	foo ~= "d";
	foo.length = 4;
	assert(foo == "codl");

	foo.length = 6;
	assert(foo == "codl\xff\xff");

	foo.length = 4;
	assert(foo == "codl");

	import std.range: retro;
	assert(equal(retro(foo[]), "ldoc"));

	import std.range: radial;
	assert(equal(radial(foo[]), "odcl"));

	import std.range: cycle;
	assert(foo[].cycle[4 .. 8].equal(foo[]));

	assert(foo[].save == foo[]);

	FixedString!10 bar;
	bar = " is nic";
	bar ~= 'e';

	immutable truth = fixedString!"codl is nice";

	assert(foo ~ bar == truth);
	assert(foo.concat!16(bar) == truth);
	assert(foo ~ bar == foo.concat!16(bar));

	FixedString!10 d;
	d = foo;
	d.length = 4;

	foreach (i, char c; d)
	{
		assert(c == foo[i]);
	}

	foo = "de";
	foo ~= "ad";
	assert(foo == "dead");
	bar = "beef";
	foo ~= bar;
	assert(foo == "deadbeef");

	assert(fixedString!"aéiou" == "aéiou");

	immutable char[4] dead = "dead";
	immutable(char)[4] beef = "beef";

	immutable FixedString!4 deader = dead;
	immutable FixedString!4 beefer = beef;
	assert(deader ~ beefer == "deadbeef");
}

///
pure @safe nothrow unittest
{
	immutable a = FixedString!16("bepis");
	assert(a.toString == "bepis");

	int[FixedString!16] table;
	table[a] = 1;

	immutable b = FixedString!16("conk");
	table[b] = 2;

	assert(table[a] == 1);
	assert(table[b] == 2);
}

/// readme example code
@safe @nogc nothrow pure unittest
{
	FixedString!14 foo = "clang";
	foo[0] = 'd';
	foo ~= " is cool";
	assert(foo == "dlang is cool");

	foo.length = 9;

	immutable bar = fixedString!"neat";
	assert(foo ~ bar == "dlang is neat");

	// wchars and dchars are also supported
	assert(FixedString!(5, wchar)("áéíóú") == "áéíóú");

	// in fact, any type is:
	immutable int[4] intArray = [1, 2, 3, 4];
	assert(FixedString!(5, int)(intArray) == intArray);
}

@system unittest
{
	import std.exception: assertThrown;
	import core.exception: AssertError;

	assertThrown!AssertError(FixedString!2("too long"));

	FixedString!2 a = "uh";
	assertThrown!AssertError(a[69] = 'a');
	assertThrown!AssertError(a.concat!1(a));
}


source/fixedstring/opapplymixin.d
module fixedstring.opapplymixin;

package(fixedstring):

string opApplyWorkaround() nothrow pure @safe
{
	// dfmt off
	return paramNumbers("") ~
	paramNumbers("@safe") ~
	paramNumbers("@nogc") ~
	paramNumbers("@safe @nogc") ~
	paramNumbers("nothrow") ~
	paramNumbers("@safe nothrow") ~
	paramNumbers("@nogc nothrow") ~
	paramNumbers("@safe @nogc nothrow") ~
	paramNumbers("pure") ~
	paramNumbers("pure @safe") ~
	paramNumbers("pure @nogc") ~
	paramNumbers("pure @safe @nogc") ~
	paramNumbers("pure nothrow") ~
	paramNumbers("pure @safe nothrow") ~
	paramNumbers("pure @nogc nothrow") ~
	paramNumbers("pure @safe @nogc nothrow");
	// dfmt on
}

private string paramNumbers(in string params) nothrow pure @safe
{
	// dfmt off
	return good(1, params, true) ~
	good(2, params, true) ~
	good(1, params, false) ~
	good(2, params, false);
	// dfmt on
}

private string delegateType(in int n, in string attributes) nothrow pure @safe
{
	string params;
	switch (n)
	{
	case 1:
		params = "ref CharT";
		break;
	case 2:
		params = "ref int, ref CharT";
		break;
	default:
		assert(false, "unsupported number of parameters.");
	}

	return "delegate(" ~ params ~ ") " ~ attributes;
}

private string good(in int n, in string parameters, in bool isConst) nothrow pure @safe
{
	string s;
	if (isConst)
	{
		s = "const";
	}
	else
	{
		s = "";
	}

	string result = "
		int opApply(scope int " ~ delegateType(n, parameters) ~ " dg) " ~ s ~ "
		{
			int result = 0;

			for (int i = 0; i != length; ++i)
			{
				CharT temp = data[i];
				" ~ resultAssign(n) ~ "
				if (result)
				{
					break;
				}
			}

			return result;
		}";

	return result;
}

private string resultAssign(in int n) nothrow pure @safe
{
	switch (n)
	{
	case 1:
		return "result = dg(temp);";
	case 2:
		return "result = dg(i, temp);";
	default:
		assert(false, "this will never happen.");
	}
}


source/fixedstring/fixedstringrange.d
module fixedstring.fixedstringrange;

package(fixedstring):

struct FixedStringRange(DataType)
{
	private const(DataType)[] source;
	private size_t startIndex;
	private size_t length_;
	size_t length()
	{
		return length_;
	}

	@disable this();

	package this(in DataType[] source)
	{
		this.source = source;
		this.length_ = source.length;
	}

	bool empty() const
	{
		return length_ == 0;
	}

	DataType front()
	in (!empty)
	{
		return source[startIndex];
	}

	void popFront()
	in (!empty)
	{
		++startIndex;
		--length_;
	}

	typeof(this) save()
	{
		return this;
	}

	DataType back()
	in (!empty)
	{
		return source[startIndex + (length_ - 1)];
	}

	void popBack()
	in (!empty)
	{
		--length_;
	}

	DataType opIndex(in size_t index)
	in (index < length_)
	{
		return source[startIndex + index];
	}

	auto opSlice(in size_t first, in size_t last)
	in (first <= length_ && last <= length_)
	{
		return typeof(this)(source[(startIndex + first) .. (startIndex + last)]);
	}
}


source/fixedstring.d
/**
* a @safe, @nogc-compatible template string type.
* Authors: Susan
* Date: 2021-11-06
* Licence: AGPL-3.0 or later
* Copyright: Susan, 2021
*/
module fixedstring.fixedstring;

import fixedstring.opapplymixin;
import std.traits: isSomeChar;

/// short syntax.
auto fixedString(string s)()
{
	import std.math.algebraic: nextPow2;
	return FixedString!(nextPow2(s.length))(s);
}

///
@safe @nogc nothrow unittest
{
	enum testString = "dlang is good";
	immutable foo = fixedString!(testString);
	immutable bar = FixedString!16(testString);

	assert(foo == bar);
}

///
struct FixedString(size_t maxSize, CharT = char)
{
	invariant (length_ <= data.length);

	enum size = maxSize; ///

	private size_t length_;
	private CharT[maxSize] data = ' ';

	///
	size_t length() const @nogc nothrow pure @safe
	{
		return length_;
	}

	/// ditto
	void length(in size_t rhs) @nogc nothrow pure @safe
	in (rhs <= maxSize)
	{
		if (rhs >= length)
		{
			data[length .. rhs] = CharT.init;
		}
		length_ = rhs;
	}

	/// constructor
	this(in CharT[] rhs)
	in (rhs.length <= maxSize)
	{
		length = rhs.length;
		data[0 .. length] = rhs[];
	}

	/// assignment
	void opAssign(in CharT[] rhs)
	in (rhs.length <= maxSize)
	{
		length = rhs.length;
		data[0 .. length] = rhs[];
	}

	/// ditto
	void opAssign(T: CharT, size_t n)(in FixedString!(n, T) rhs)
	in (rhs.length <= maxSize)
	{
		length = rhs.length;
		data[0 .. length] = rhs[0 .. length];
	}

	/// ditto
	void opOpAssign(string op)(in CharT[] rhs)
	if (op == "~")
	in (length + rhs.length <= maxSize)
	{
		immutable oldLength = length;
		length = length + rhs.length;
		data[oldLength .. length] = rhs[];
	}

	/// ditto
	void opOpAssign(string op)(in CharT rhs)
	if (op == "~")
	in (length + 1 <= maxSize)
	{
		length = length + 1;
		data[length - 1] = rhs;
	}

	/// ditto
	void opOpAssign(string op, T: CharT, size_t n)(in FixedString!(n, T) rhs)
	if (op == "~")
	in (length + rhs.length <= maxSize)
	{
		immutable oldLength = length;
		length = length + rhs.length;
		data[oldLength .. length] = rhs[0 .. rhs.length];
	}

	/// array features...
	auto opSlice(in size_t first, in size_t last) const @nogc nothrow pure @safe
	in (first <= length && last <= length)
	{
		return data[first .. last];
	}

	/// ditto
	size_t opDollar(size_t pos)() const @nogc nothrow pure @safe
	if (pos == 0)
	{
		return length;
	}

	/// ditto
	auto opIndex() const
	{
		return FixedStringRangeInterface!CharT(data[0 .. length]);
	}

	/// ditto
	CharT opIndex(in size_t index) const @nogc nothrow pure @safe
	in (index < length)
	{
		return data[index];
	}

	/// ditto
	void opIndexAssign(in CharT rhs, in size_t index)
	in (index < length)
	{
		data[index] = rhs;
	}

	/// equality
	bool opEquals(T : CharT, size_t n)(in FixedString!(n, T) rhs) const
	{
		import std.algorithm.comparison: equal;
		return this[].equal(rhs[]);
	}

	/// ditto
	bool opEquals(in CharT[] s) const
	{
		if (length != s.length)
		{
			return false;
		}

		import std.algorithm.comparison: equal;
		return this[].equal(s[]);
	}

	/// concatenation. note that you should probably use the ~ operator instead - only use this version when you are pressed for ram and aren't making many calls, or you will end up with template bloat.
	auto concat(size_t s, T: CharT, size_t n)(in FixedString!(n, T) rhs) const
	in (s >= length + rhs.length)
	{
		FixedString!(s) result;

		result = this;
		result ~= rhs;

		return result;
	}

	/// concatenation operator. generally, you should prefer this version.
	auto opBinary(string op, T: CharT, size_t n)(in FixedString!(n, T) rhs) const
	if (op == "~")
	{
		import std.math.algebraic: nextPow2;
		return concat!(nextPow2(size + rhs.size))(rhs);
	}

	static if (isSomeChar!CharT)
	{
		///
		const(CharT)[] toString() const nothrow pure @safe
		{
			return data[0 .. length].idup;
		}
	}

	///
	size_t toHash() const @nogc nothrow pure @safe
	{
		ulong result = length;
		foreach (CharT c; data[0 .. length])
		{
			result += c;
		}
		return result;
	}

	mixin(opApplyWorkaround);
}

/// readme example code
@safe @nogc nothrow pure unittest
{
	FixedString!14 foo = "clang";
	foo[0] = 'd';
	foo ~= " is cool";
	assert(foo == "dlang is cool");

	foo.length = 9;

	immutable bar = fixedString!"neat";
	assert(foo ~ bar == "dlang is neat");

	// wchars and dchars are also supported
	assert(FixedString!(5, wchar)("áéíóú") == "áéíóú");

	// in fact, any type is:
	immutable int[4] intArray = [1, 2, 3, 4];
	assert(FixedString!(5, int)(intArray) == intArray);
}

private struct FixedStringRangeInterface(DataType)
{
	private const(DataType)[] source;
	private size_t startIndex;
	private size_t length_;
	size_t length()
	{
		return length_;
	}

	@disable this();

	package this(in DataType[] source)
	{
		this.source = source;
		this.length_ = source.length;
	}

	bool empty() const
	{
		return length_ == 0;
	}

	DataType front()
	in (!empty)
	{
		return source[startIndex];
	}

	void popFront()
	in (!empty)
	{
		++startIndex;
		--length_;
	}

	typeof(this) save()
	{
		return this;
	}

	DataType back()
	in (!empty)
	{
		return source[startIndex + (length_ - 1)];
	}

	void popBack()
	in (!empty)
	{
		--length_;
	}

	DataType opIndex(in size_t index)
	in (index < length_)
	{
		return source[startIndex + index];
	}

	auto opSlice(in size_t first, in size_t last)
	in (first <= length_ && last <= length_)
	{
		return typeof(this)(source[(startIndex + first) .. (startIndex + last)]);
	}
}

@safe @nogc nothrow pure unittest
{
	immutable string temp = "cool";
	auto a = FixedString!8(temp);
	assert(a[0] == 'c');
	assert(a == "cool");

	import std.algorithm.comparison: equal;
	assert(a[].equal("cool"));
	assert(a[0 .. $] == "cool");

	a[2] = 'd';
	assert(a == "codl");
	assert(a != "");

	a ~= "d";
	a.length = 4;
	assert(a == "codl");

	a.length = 6;
	assert(a == "codl\xff\xff");

	a.length = 4;
	assert(a == "codl");

	import std.range: retro;
	assert(equal(retro(a[]), "ldoc"));

	import std.range: radial;
	assert(equal(radial(a[]), "odcl"));

	import std.range: cycle;
	assert(a[].cycle[4 .. 8].equal(a[]));

	assert(a[].save == a[]);

	FixedString!10 b;
	b = " is nic";
	b ~= 'e';
	assert(a ~ b == "codl is nice");
	assert(a.concat!16(b) == "codl is nice");
	assert(a ~ b == a.concat!16(b));

	FixedString!10 d;
	d = a;

	foreach (i, char c; a)
	{
		switch (i)
		{
		case 0:
			assert(c == 'c');
			break;

		case 1:
			assert(c == 'o');
			break;

		case 2:
			assert(c == 'd');
			break;

		case 3:
			assert(c == 'l');
			break;

		default:
			assert(false);
		}
	}

	a = "de";
	a ~= "ad";
	assert(a == "dead");
	b = "beef";
	a ~= b;
	assert(a == "deadbeef");

	assert(fixedString!"aéiou" == "aéiou");

	immutable char[4] dead = "dead";
	immutable(char)[4] beef = "beef";

	immutable FixedString!4 deader = dead;
	immutable FixedString!4 beefer = beef;
	assert(deader ~ beefer == "deadbeef");
}

pure @safe nothrow unittest
{
	immutable a = FixedString!16("bepis");
	assert(a.toString == "bepis");

	int[FixedString!16] table;
	table[a] = 1;

	immutable b = FixedString!16("conk");
	table[b] = 2;

	assert(table[a] == 1);
	assert(table[b] == 2);
}

@system unittest
{
	import std.exception: assertThrown;
	import core.exception: AssertError;

	assertThrown!AssertError(FixedString!2("too long"));

	FixedString!2 a = "uh";
	assertThrown!AssertError(a[69] = 'a');
	assertThrown!AssertError(a.concat!1(a));
}
